plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java

```diff
@@ -217,12 +218,33 @@ InputStream readBlob(String blobName, long position, long length) throws IOExcep
 
     /**
      * Writes a blob in the specific bucket
-     *  @param inputStream content of the blob to be written
+     * @param bytes       content of the blob to be written
+     * @param failIfAlreadyExists whether to throw a FileAlreadyExistsException if the given blob already exists
+     */
+    void writeBlob(String blobName, BytesReference bytes, boolean failIfAlreadyExists) throws IOException {
+        if (bytes.length() > getLargeBlobThresholdInBytes()) {
+            // Compute crc32c here so #writeBlobResumable forces the integrity check on the resumable upload.
+            // This is needed since we rely on atomic write behavior when writing BytesReferences in BlobStoreRepository which is not
+            // guaranteed for resumable uploads.
+            writeBlobResumable(
+                BlobInfo.newBuilder(bucketName, blobName).setCrc32c(Crc32C.checksum(bytes)).build(),
+                bytes.streamInput(), bytes.length(), failIfAlreadyExists);
+        } else {
+            writeBlob(blobName, bytes.streamInput(), bytes.length(), failIfAlreadyExists);
```

Comment:
No need to do any manual shenanigans here, the normal non-resumable upload path will is not affected by the issue since it always sets checksum validation on the upload.